### PR TITLE
Groups

### PR DESCRIPTION
**Pull request by roed314**
Add finite groups.  To test, explore https://groups.lmfdb.xyz/Groups/Abstract.  Some features to look at
* Currently have data up to order 511; planning to add more.
* Basic and advanced search (like abelian varieties).
* Can [search for subgroups](https://groups.lmfdb.xyz/Groups/Abstract/?search_type=Subgroups).
* There are [statistics](https://groups.lmfdb.xyz/Groups/Abstract/stats) and [interesting groups](https://groups.lmfdb.xyz/Groups/Abstract/interesting) pages
* There are rcs knowls
* There are tests (though more would always be nice)
* We describe boolean properties using a string format that gives some of the implications
* There are images for individual groups, based on sizes and orders of conjugacy classes
* The constructions section lists a polycyclic presentation (solvable groups) or a set of generating permutations (nonsolvable groups), as well as decompositions as direct/semidirect/nonsplit products.
* There are several views for investigating subgroups.  In addition to the subgroup search, we also show subgroup lattices and lists of subgroups up to abstract isomorphism.  Subgroups are computed up to a mix of conjugacy and automorphism, depending on how many there are.
* There are complex and rational character tables.

### CODE REVIEW

**Comment by AndrewVSutherland** [comment]
I plan to take a look at this tomorrow (and merge if all goes well).  I'd encourage anyone and everyone to take a look.

**Comment by roed314** [comment]
In terms of the sidebar, one thing I'll note is that we eventually hope to include abstract groups that are not part of the SmallGroups database.  So I don't know if `Small Groups` is the best label to include in the sidebar.  Would `Abstract Groups` be better?  Does anyone have other ideas?

File: lmfdb/groups/abstract/web_groups.py [hunk 1]
```diff
@@ -0,0 +1,1223 @@
...
+            for k, v in self.conjugacy_class_divisions.items()
+        ]
+        ccdivs.sort(key=lambda x: x["classes"][0].counter)
+        return ccdivs
+
+    @lazy_attribute
+    def cc_to_div(self):
+        # Need to map cc's to their divisions
+        ctor = {}
+        for k, vs in self.conjugacy_class_divisions.items():
```
Comment (edgarcosta) [inline lmfdb/groups/abstract/web_groups.py:279]: isn't self.conjugacy_class_divisions a list?

File: lmfdb/groups/abstract/web_groups.py [hunk 2]
```diff
@@ -0,0 +1,1223 @@
...
+                labels.append(make_full(label))
+        for llist in [self.complements, self.contained_in, self.contains]:
+            if llist:
+                labels.extend([make_full(label) for label in llist])
+        return list(db.gps_subgroups.search({"label": {"$in": labels}}))
+
+    def autjugate_subgroups(self):
+        return [
+            H
+            for H in self.amb.subgroups.values()
```
Comment (edgarcosta) [inline lmfdb/groups/abstract/web_groups.py:1064]: I found:
- self.amb
- self.ambient
- self.ambient_gp
very confusing.

File: lmfdb/groups/abstract/web_groups.py [hunk 3]
```diff
@@ -0,0 +1,1223 @@
...
+    def autjugacy_classes(self):
+        cl = self.conjugacy_classes  # creates autjugacy classes
+        assert cl
+        return self.autjugacy_classes
+
+    @lazy_attribute
+    def sorted_cc_divisions(self):
+        ccdivs = [
+            {"label": k, "classes": v}
+            for k, v in self.conjugacy_class_divisions.items()
```
[the remark this answers is not in the document]
Comment (roed314) [reply lmfdb/groups/abstract/web_groups.py]: This is no longer used; I just deleted it.

**Comment by roed314** [comment]
@jwj61 I've removed the things like "A-group but not a Z-group" and moved them into knowls.

**Pull request merged**